api/src/wfl/module/covid.clj

```diff
@@ -23,6 +24,61 @@
 (defn create-covid-workload!
   [tx {:keys [source sink executor] :as request}])
 
+(defn start-covid-workload!
+  [{:keys [submission-id workspace] :as workload}])
+
+(defn ^:private snapshot-error-map
+  "Return default ExceptionInfo map for snapshot-related exceptions."
+  [{:keys [items sink] :as _workload} {:keys [id snapshot_id] :as _record}]
+  {:workspace (:workspace sink)
+   :table items
+   :id id
+   :snapshot_id snapshot_id})
+
+(defn ^:private try-get-snapshot-reference
+  "Attempt to fetch snapshot reference with id SNAPSHOT_REFERENCE_ID
+  from SINK-derived workspace."
+  [{:keys [sink] :as workload} {:keys [snapshot_reference_id] :as record}]
+  (let [workspace (:workspace sink)]
+    (try
+      (rawls/get-snapshot-reference workspace snapshot_reference_id)
+      (catch Throwable cause
+        (throw (ex-info "Rawls unable to fetch snapshot reference"
+                        (util/deep-merge
+                         (snapshot-error-map workload record)
+                         {:snapshot_reference_id snapshot_reference_id})
+                        cause))))))
+
+(defn ^:private snapshot-imported?
+  "Check if snapshot with id SNAPSHOT_ID in TDR has been imported to
+  SINK-derived workspace and can be successfully fetched."
+  [workload {:keys [snapshot_id snapshot_reference_id] :as record}]
+  {:pre [(some? snapshot_id)]}
+  (and (some? snapshot_reference_id)
+       (some? (try-get-snapshot-reference workload record))))
+
+(defn ^:private import-snapshot!
+  "Import snapshot with id SNAPSHOT_ID from TDR to workspace.
+  Use transaction TX to update ITEMS table with resulting reference id."
+  [tx {:keys [items sink] :as workload} {:keys [id snapshot_id] :as record}]
+  (let [workspace (:workspace sink)
+        name (util/randomize "placeholder-snapshot-ref-name")
+        reference-id (rawls/create-snapshot-reference workspace snapshot_id name)]
```

Comment:
If this chokes, I think that an error is thrown and we never get to our more detailed thrown error below.

Is there a Clojure one-liner like Scala's [`Try… getOrElse…`](https://stackoverflow.com/questions/40212369/try-and-getorelse-scala/40212461) that can handle thrown exceptions in assignment gracefully?  Or a better way?

Reply:
Nope - you can write one though:
```clojure
(defn on-error [thunk handler]
  (try
    (thunk)
    (catch Throwable t
      (handler t)))))

(let [reference-id (on-error #(create-snapshot-reference ...) #(add-error-context context %))]
  ..)
````

Reply:
Clojurists generally just `try` like savages.

Reply:
I ended up getting what I needed from `util/do-or-nil`.